docs/architecture/adr-008-chain-endpoint-handle-consolidation.md

```diff
@@ -0,0 +1,123 @@
+# ADR 008: ChainEndpoint/ChainHandle Consolidation
+
+## Changelog
+* 2022-04-19: Initial Proposal
+
+## Context
+There is a lot of common methods in the `ChainHandle` and `ChainEndpoint` traits, sometimes with minute differences between one another. This document provides a way to remove the duplication of methods for increased maintainability of the codebase, along with a few suggestions to standardize the method signatures. It also suggests a procedural macro to simplify the currently repetitive implementation of the `query_*` methods, which would be made possible given the proposed standardization of the method signatures.
+
+## Decision
+
+### Query methods parameters
+There are currently discrepancies between how methods take their arguments. Some take a `request` object, and others take fine-grained arguments that will be used to build a `request` object in the implementation of the method. For example, `query_consensus_state()` takes arguments that will be used to build a request object, whereas `query_consensus_states()` takes a request object directly.
+```rust
+fn query_consensus_state(
+    &self,
+    client_id: ClientId,
+    consensus_height: Height,
+    query_height: Height,
+) -> ...;
+
+fn query_consensus_states(
+    &self,
+    request: QueryConsensusStatesRequest,
+) -> ...;
+```
+
+All methods will be refactored to take a request object as argument.
```

Comment:
I would rather go the other way and have all methods take multiple parameters. Reason for this is that the `*Request` datatypes are Protobuf definitions which I would rather not expose in the public API directly, as those are merely an internal implementation detail.